app/com/linkedin/drelephant/tuning/engine/SparkExecutionEngine.java

```diff
@@ -0,0 +1,85 @@
+package com.linkedin.drelephant.tuning.engine;
+
+import java.util.HashMap;
+import java.util.List;
+import models.AppResult;
+import models.JobExecution;
+import models.JobSuggestedParamSet;
+import models.JobSuggestedParamValue;
+import models.TuningAlgorithm;
+import models.TuningJobDefinition;
+import models.TuningParameter;
+import org.apache.log4j.Logger;
+import com.avaje.ebean.Expr;
+import com.avaje.ebean.ExpressionList;
+import com.linkedin.drelephant.tuning.ExecutionEngine;
+
+
+/**
+ * This class represents Spark Exectuion Engine. It handles all the cases releated to Spark Engine
+ */
+public class SparkExecutionEngine implements ExecutionEngine {
+  private final Logger logger = Logger.getLogger(getClass());
+  private String executionEngineName = "SPARK";
+
+  public String getExecutionEngineName() {
+    return this.executionEngineName;
+  }
+
+  @Override
+  public void computeValuesOfDerivedConfigurationParameters(List<TuningParameter> derivedParameterList,
+      List<JobSuggestedParamValue> jobSuggestedParamValue) {
+
+  }
+
+  @Override
+  public ExpressionList<JobSuggestedParamSet> getPendingJobs() {
+    return JobSuggestedParamSet.find
+        .select("*")
+        .fetch(JobSuggestedParamSet.TABLE.jobDefinition, "*")
+        .where()
+        .or(Expr.or(Expr.eq(JobSuggestedParamSet.TABLE.paramSetState, JobSuggestedParamSet.ParamSetStatus.CREATED),
+            Expr.eq(JobSuggestedParamSet.TABLE.paramSetState, JobSuggestedParamSet.ParamSetStatus.SENT)),
+            Expr.eq(JobSuggestedParamSet.TABLE.paramSetState, JobSuggestedParamSet.ParamSetStatus.EXECUTED))
+       // .eq(JobSuggestedParamSet.TABLE.isParamSetDefault, 0)
+        .eq(JobSuggestedParamSet.TABLE.tuningAlgorithm + "." + TuningAlgorithm.TABLE.jobType,
```

Comment:
Here, the same DB query will be executed twice i.e. once for HBT and once again for OBT. Can't we fetch all the jobs for HBT/OBT one time and then based on job type make calls to specific execution engine implementations? This will reduce the DB queries made. Also I do not really see any benefit for giving flexibility to developers to have custom implementations to fetch pending jobs by job type. We won't have TuneIn tables based on job types or have different ways for fetching data from backend tables. Infact it can be argued that the developer adding new execution engine or the tuning type need not know internal details of the EBean layer and the backing DB tables at all. Execution engine implementation can provide a getJobType method which can then be used to make relevant calls to bean.
Same comment applies to MRExecutionEngine.

Reply:
I think we can have one DB query to even fetch all the rows (Even for all the four modules , and then  each class can take care of it ) . This will reduce the DB queries drastically ,but then all the module will depend on the DB query module . I will start abstracting out DB queries . But will have more discussion about the queries all rows vs each individual module takes care of their own data

Reply:
https://jira01.corp.linkedin.com:8443/browse/DSS-18339